WIP : Msi stock indexing #48

larsroettig opened the pull request.

Commit:
MSI magento-engcom/magento#48:
- Add Setup for index table
- Add new Interface"

Commit:
MSI magento-engcom/magento#48:
- Refactoring to new Setup structure
- Implement SourceItem add entry to di.xml
- WIP Register Basic Indexer SourceItem

Commit:
MSI magento-engcom/magento#48:
- Impelemnt basic import funktion

TODOS:
- Add some bunching function for index
- Add Mysql Index to  inventory_stock_item_index table
- Documentation and clean up

larsroettig commented:
@maghamed hi it is only basic version I think must also implement a bunching but the code is working some can be optimized. Have you a good example how can I test it?

Commit:
Add bugfix for tests

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/AbstractAction.php:
```diff
@@ -0,0 +1,150 @@
+use \Psr\Log\LoggerInterface;
+use \Magento\Framework\App\ResourceConnection;
+
+abstract class AbstractAction
```
Please don't introduce AbstractAction class, as Magento doesn't recommend to use Inheritance in newly created code.

Create a dedicated indexer model which would be injected via DI into each of the class:
```php
Row
Rows
Full
```

and make each of these classes to proxy call to that indexer model

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/AbstractAction.php:
```diff
@@ -0,0 +1,150 @@
+    public function __construct(ResourceConnection $resource, LoggerInterface $logger)
+    {
+        $this->resource = $resource;
+        $this->connection = $resource->getConnection();
```
it's not recommended to make anything except assignment in the constructors.
Like in your case where you do 
` $this->connection = $resource->getConnection();`

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem.php:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Stock item Indexer @todo
+ */
+class StockItem implements \Magento\Framework\Indexer\ActionInterface, \Magento\Framework\Mview\ActionInterface
```
It's wrong to Make indexer implementing both of these interfaces, because these interfaces responsible for different things. Doing so we violate Single Responsibility Principle. 

We need to have two different classes, each one implementing own interface.

Like in **CatalogSearch** module
https://github.com/magento/magento2/blob/develop/app/code/Magento/CatalogSearch/Model/Indexer/Mview/Action.php
implements \Magento\Framework\Mview\ActionInterface
which is declared in - https://github.com/magento/magento2/blob/develop/app/code/Magento/CatalogSearch/etc/mview.xml#L9


and Magento\CatalogSearch\Model\Indexer\Fulltext implements  only \Magento\Framework\Indexer\ActionInterface
which is declared in 
https://github.com/magento/magento2/blob/develop/app/code/Magento/CatalogSearch/etc/indexer.xml#L9

Let's do the same way

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/AbstractAction.php:
```diff
@@ -0,0 +1,150 @@
+            $indexItems = $this->fetchIndexItems($sourceItems);
+            if (is_array($indexItems) && count($indexItems) > 0) {
+                $this->cleanIndexTable($sourceItems);
+                $this->updateIndexTable($indexItems);
```
It's better to introduce implementation of IndexerInterface
https://github.com/magento/magento2/blob/develop/lib/internal/Magento/Framework/Indexer/SaveHandler/IndexerInterface.php

and use it for the operations of clean and delete index

maghamed commented on app/code/Magento/Inventory/Setup/Operation/CreateStockItemIndexTable.php:
```diff
@@ -0,0 +1,88 @@
+
+use Magento\InventoryApi\Api\Data\StockItemInterface;
+
+class CreateStockItemIndexTable
```
This should be implemented similarly to what we have in CatalogSearch 
where
`Magento\CatalogSearch\Model\Indexer\IndexStructure`
responsible for creating StockItem index structure.

IndexStructure implements interface 
`Magento\Framework\Indexer\IndexStructureInterface`

we will reuse this class for index structure creation because we will create dedicated index table for each particular Stock (stock_id). To make the system being scalable horizontally.
As each index table will hold StockItems for each particular Scope (SalesChannel).

For example, having 5 websites each of which represents Sales Channel we will end up with 5 index tables:
```
inventory_stock_item_index_1 - for website_id 1 
inventory_stock_item_index_2 - for website_id 2
inventory_stock_item_index_3 - for website_id 3
inventory_stock_item_index_4 - for website_id 4
inventory_stock_item_index_5 - for website_id 5
```

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -53,13 +59,15 @@ public function __construct(
         CreateStockTable $createStockTable,
-        CreateStockSourceLinkTable $createStockSourceLinkTable
+        CreateStockSourceLinkTable $createStockSourceLinkTable,
+        CreateStockItemIndexTable $createStockItemIndexTable
```
No need to introduce index table during the installation , it should be done at the time of re-indexation

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/Action/Rows.php:
```diff
@@ -0,0 +1,34 @@
+        try {
+            $this->reindexRows($ids);
+        } catch (\Exception $e) {
+            throw new \Magento\Framework\Exception\LocalizedException(__($e->getMessage()), $e);
```
it's not correct translation, 
`__($e->getMessage()), $e);`
Because \Exception $e ,so it's not a successor of Localized Exception, thus $e->getMessage()
could return a string for which you don't have precise translation

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/Action/Row.php:
```diff
@@ -0,0 +1,34 @@
+        try {
+            $this->reindexRows([$id]);
+        } catch (\Exception $e) {
+            throw new \Magento\Framework\Exception\LocalizedException(__($e->getMessage()), $e);
```
it's not correct translation,
`__($e->getMessage()), $e);`
Because \Exception $e ,so it's not a successor of Localized Exception, thus $e->getMessage()
could return a string for which you don't have precise translation

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/Action/Row.php:
```diff
@@ -0,0 +1,34 @@
+     *
+     * @return void
+     */
+    public function execute($id = null)
```
why do we allow to accept NULL value if we throw an exception if NULL passed as a parameter?

maghamed commented on app/code/Magento/InventoryApi/Api/Data/StockItemInterface.php:
```diff
@@ -0,0 +1,106 @@
+ * Interface StockItemInterface
+ * @api
+ */
+interface StockItemInterface extends ExtensibleDataInterface
```
For now we don't have a Service API to retrieve Stock Items **for a given Stock_ID**

That's why for now we can't get StockItems in any case
`
These API should be introduced to InventoryApi\Api\StockItemManagerInterface`

maghamed commented on app/code/Magento/Inventory/etc/indexer.xml:
```diff
@@ -0,0 +1,13 @@
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Indexer/etc/indexer.xsd">
+    <indexer id="inventory_stock_item" view_id="inventory_stock_item" class="Magento\Inventory\Model\Indexer\StockItem">
+        <title translate="true">StockItem</title>
+        <description translate="true">Index stock item</description>
```
Please provide more clear description

maghamed reviewed (changes requested):
For Big Merchants, who have a lot of SKUs and a lot of Sales channels (scopes in which products would be sold). We want to prevent a multiplication of data in the index table : 
**SKUs Qty * Number of Sales Channels**

Because that could introduce performance impact both on Read and Write index operations.

That's why it's proposed to split indexes by the Scope.
Taking into account that we have 1 - 1 relation between  Scope (Sales Channel) and Stock.
We could Build independent index table per each Stock (Stock_id).

Stock Id could be used as Dimension which will help us to resolve correct index we need to refer to get Product Qty.

For example, having 5 websites with ids (1 .. 5)
we will end up creating 5 different index tables (with the same structure):
```
inventory_stock_item_stock_1
inventory_stock_item_stock_2
inventory_stock_item_stock_3
inventory_stock_item_stock_4
inventory_stock_item_stock_5
```

You can refer for similar implementation in the CatalogSearch Index
`Magento\CatalogSearch\Model\Indexer\Fulltext`
Also, we have a requirement to have index aliasing during the full reindex happened, so we need to have switchable indexes to make Front-End responsive while Full Reindex happen.


Also, please take into account that when the SourceItem data change -> **we need to upgrade Stock Item Indexes for each of the Stock to which given Source was assigned**.
If Source was assigned to 3 stocks, for example, we need to refresh data in the 3 StockItem indexes

Looks like, for this operation we need to introduce new service
`Magento\InventoryApi\Api\GetAssignedStocksForSourceInterface::execute($sourceId);`

Commit:
Remove setup

Commit:
MSI magento-engcom/magento#48:
- Clean Up and implement basic structure

larsroettig commented:
Hi @maghamed,
i see some Problems first from all the must logic to switch an Index with zero down time is in the
**CatalogSearch\Model\Indexer**  Module.

But is it very useful Core and PluginDevleoper to Register a  zero down time indexer. 
My suggestion is move **IndexSwitcherProxy** and  **IndexStructureFactory** to the 
**Magento\Indexer** and mark the clases  as Deprecte. 

Magento\CatalogSearch\Model\Indexer\IndexSwitcherProxy
Magento\CatalogSearch\Model\Indexer\IndexStructureFactory

Commit:
Add basic structure

Commit:
Implement mview

naydav commented on app/code/Magento/Inventory/Model/GetAssignedStocksForSource.php:
```diff
@@ -0,0 +1,111 @@
+    private function getAssignedStockIds($sourceId)
+    {
+        $searchCriteria = $this->searchCriteriaBuilder
+            ->addFilter(StockSourceLink::SOURCE_ID, (int)$sourceId)
```
Cast to int is extra

naydav commented on app/code/Magento/Inventory/Model/GetAssignedStocksForSource.php:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * Get all linked StockIds by given sourceId.
+     *
+     * @param $sourceId
```
Missed parameter type

naydav commented on app/code/Magento/Inventory/Model/Indexer/StockItem.php:
```diff
@@ -0,0 +1,54 @@
+use Magento\Inventory\Model\Indexer\StockItem\IndexStructure;
+
+/**
+ * Stock item Indexer @todo
```
Pls delete or add more description to @todo annotation

naydav commented on app/code/Magento/Inventory/Model/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,167 @@
+        Batch $batch,
+        ResourceConnection $resource,
+        array $data,
+        $batchSize = 100
```
Move default value to di configuration

maghamed replied:
use DI Type configuration for batchSize, especially taking into account that you will use Type for IndexStructure

naydav commented on app/code/Magento/Inventory/Model/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,167 @@
+    /**
+     * @var array
+     */
+    private $data;
```
Need to more clear name?
What is mean data?

naydav commented on app/code/Magento/Inventory/Model/StockItemManager.php:
```diff
@@ -0,0 +1,112 @@
+ *
+ * @codeCoverageIgnore
+ */
+class StockItemManager extends AbstractExtensibleModel implements StockItemManagerInterface
```
As for me *Manager is not good name for "entity"
Maybe it is better simple "StockItem"

But pay attention we already have **\Magento\InventoryApi\Api\Data\StockItemInterface**

maghamed replied:
why do we have StockItemManager which behaves like DataInterface,

**Manager** means to make some actions over some data.

naydav commented on app/code/Magento/InventoryApi/Api/Data/StockItemManagerInterface.php:
```diff
@@ -0,0 +1,106 @@
+    public function setStockId($stockId);
+    
+    /**
+     * Get stock item sku
```
it is not stock item sku

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,167 @@
+     * @var string
+     */
+    const DATA_KEY_INDEXER_ID = 'indexer_id';
+    const DATA_KEY_INDEXER_SUFFIX = 'indexer_suffix';
```
no need to make these fields as constants, because constants are the contract which could be accessible outside of the class (like public methods)

it's better to make them **private static** in this case, so, no one from outside could rely on these values.
And we will not duplicate these data for all instantiated classes.

naydav commented on app/code/Magento/InventoryApi/Api/GetAssignedStocksForSourceInterface.php:
```diff
@@ -0,0 +1,26 @@
+interface GetAssignedStocksForSourceInterface
+{
+    /**
+     * Get Stock assigned to Sources
```
visa versa
Get Stocks assigned to Source

naydav commented on app/code/Magento/Inventory/etc/mview.xml:
```diff
@@ -0,0 +1,15 @@
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Mview/etc/mview.xsd">
+    <view id="inventory_stock_item" class="Magento\Inventory\Model\Indexer\Mview\StockItemAction" group="indexer">
+        <subscriptions>
+            <table name="inventory_source_item" entity_column="source_id" />
```
Do we need to process disabling/enabling of products?

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,167 @@
+     * @param int $batchSize #
+     */
+    public function __construct(
+        IndexStructure $indexStructure,
```
it's better to use IndexStructureinterface here and specify in DI Type what exact implementation should use here

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem.php:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Indexer ID in configuration
+     */
+    const INDEXER_ID = 'inventory_stock_item_index';
```
hm... it's better to keep this constant away from the concrete implementation, but for now I don't see a place where we could move it

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,167 @@
+    private function getIndexSuffix()
+    {
+        if (isset($this->data[self::DATA_KEY_INDEXER_SUFFIX])) {
+            return $this->data[self::DATA_KEY_INDEXER_SUFFIX];
```
how can we set these data?

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem.php:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Stock item Indexer @todo
+ */
+class StockItem implements \Magento\Framework\Indexer\ActionInterface
```
let's introduce 
```php
interface StockItemIndexerInterface extends \Magento\Framework\Indexer\ActionInterface
{
    /**
      * Indexer ID in configuration
      */
     const INDEXER_ID = 'inventory_stock_item_index';
}
```
and make this class to implement this interface.

But Ideally to have getName method in ActionInterface which returned value initialized from the configuration, which we can't add because of BackwardCompatibility

maghamed commented on app/code/Magento/InventoryApi/Api/Data/StockItemManagerInterface.php:
```diff
@@ -0,0 +1,106 @@
+ * Interface StockItemManagerInterface
+ * @api
+ */
+interface StockItemManagerInterface extends ExtensibleDataInterface
```
why do we need this DataInterface? 
it duplicated StockItemInterface, but called manager and never used

looks like a mistake

maghamed reviewed (changes requested):
In general everything good, @larsroettig please keep going

Commit:
WIP

Commit:
wip

Commit:
MSI magento-engcom/magento#48:
- Impelemnt basic import function

TODOS:
- Index switch from tmp to idx table
- Documentation and clean up

Commit:
MSI magento-engcom/magento#48:
- Add Setup for index table
- Add new Interface

I magento-engcom/magento#48:
- Add Setup for index table
- Add new Interface

Commit:
MSI magento-engcom/magento#48:
- Implement switch for Index

TODO:
- Clean Up and more Doc
- Discuss the missing website to stock mapping

ishakhsuvarov commented on app/code/Magento/Inventory/Model/Indexer/StockItem.php:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * StockItem constructor.
+     * @param DimensionFactory $dimensionFactory
+     * @param StoreManagerInterface $storeManager
```
Please update doc block.

ishakhsuvarov commented on app/code/Magento/Inventory/Model/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,79 @@
+{
+
+    const TABLE_NAME_SOURCE_ITEM = 'inventory_source_item';
+    const TABLE_NAME_STOCK_SOURCE_LINK = 'inventory_source_stock_link';
```
Please describe constants.
More info is available [in the devdocs](http://devdocs.magento.com/guides/v2.2/coding-standards/docblock-standard-general.html#constants)

ishakhsuvarov commented on app/code/Magento/Inventory/Model/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,79 @@
+     * @param \Magento\Framework\DB\Adapter\AdapterInterface $connection
+     * @return string
+     */
+    private function getSourceItemTableName($connection)
```
Please add a type hint

ishakhsuvarov commented on app/code/Magento/Inventory/Model/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,79 @@
+     * @param \Magento\Framework\DB\Adapter\AdapterInterface $connection
+     * @return string
+     */
+    private function getLinkTableName($connection)
```
Please add a type hint

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem.php:
```diff
@@ -0,0 +1,108 @@
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Inventory\Model\Indexer;
```
It's better to have a separate namespace for Indexer. Not to put all indexers to 
` Magento\Inventory\Model\*`

Let's create dedicated namespace like 

`Magento\Inventory\Indexer\`

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem.php:
```diff
@@ -0,0 +1,108 @@
+        $storeIds = array_keys($this->storeManager->getStores());
+
+        foreach ($storeIds as $storeId) {
+            $dimension = [$this->dimensionFactory->create(['name' => 'scope', 'value' => $storeId])];
```
In the documentation  https://github.com/magento-engcom/magento2/wiki/New-Indexer-for-StockItems#index-dimensions stated 

 
>That's why it's proposed to split indexes by the Scope. Taking into account that we have 1 - 1 relation between Scope (Sales Channel) and Stock 
For example, having 5 websites with ids (1 .. 5) we will end up creating 5 different index tables (with the same structure):
```
inventory_stock_item_stock_1
inventory_stock_item_stock_2
inventory_stock_item_stock_3
inventory_stock_item_stock_4
inventory_stock_item_stock_5
```

Current code where dimension is StoreId is copy-pasted from Search Indexer

maghamed commented on app/code/Magento/Inventory/Model/Indexer/StockItem/Dimension.php:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Stock Item Dimension
+ */
+class Dimension extends \Magento\Framework\Search\AbstractKeyValuePair
```
we already have 
\Magento\Framework\Search\Request\Dimension
why do we create own Dimension object?

Commit:
wip

Commit:
MSI magento-engcom/magento#48:
- Impelemnt Basic intigration test fixtures

Commit:
MSI magento-engcom/magento#48:
- Bugfix for di.xml
- Add basic call for reindex full

Commit:
MSI magento-engcom/magento#48:
- Basic Indexer with switch for tmp and normal index

TODO:
- Refactoring to remove the state folder from the module
- Doc blocks and clean up
- Add assertions to the test case

naydav commented on app/code/Magento/Inventory/Indexer/Scope/ScopeProxy.php:
```diff
@@ -0,0 +1,76 @@
+ * depending on current scope state
+ * @todo refactoring it copy from catalog search module
+ */
+class ScopeProxy implements \Magento\Framework\Search\Request\IndexScopeResolverInterface
```
Pls provide more clear class name
class name is proxy, class  interface is resolver, due to code it is factory... it is not clear what is responsibility of this class

naydav commented on app/code/Magento/Inventory/Indexer/Scope/ScopeProxy.php:
```diff
@@ -0,0 +1,76 @@
+        if (!array_key_exists($state, $this->states)) {
+            throw new UnknownStateException(__("Requested resolver for unknown indexer state: $state"));
+        }
+        return $this->objectManager->create($this->states[$state]);
```
Do we need to create a new object each time?

naydav commented on app/code/Magento/Inventory/Indexer/Scope/ScopeProxy.php:
```diff
@@ -0,0 +1,76 @@
+    private function create($state)
+    {
+        if (!array_key_exists($state, $this->states)) {
+            throw new UnknownStateException(__("Requested resolver for unknown indexer state: $state"));
```
This logic is better to place in State object. In this way, we don't provide possibility to set the wrong state

But if we look code of State object, we can not find possibility to set the wrong state.
In State objectwWe have only `useTemporaryIndex` and `useRegularIndex` methods, thus we can not set state different from these two. 
So this checking looks like redundant

maghamed replied:
agree with @naydav in the scope of UnknownStateException I described the idea of State Machine, so don't think that we need to check the state and have a factory for state creation

naydav commented on app/code/Magento/Inventory/Indexer/Scope/TemporaryResolver.php:
```diff
@@ -0,0 +1,43 @@
+ * Resolves name of a temporary table for indexation
+ * @todo refactoring it copy from catalog search module
+ */
+class TemporaryResolver implements \Magento\Framework\Search\Request\IndexScopeResolverInterface
```
I understand that is taken from search index but I don't like naming (TemporaryResolver, IndexScopeResolver, ScopeProxy)
@maghamed what do you think about this naming?

maghamed replied:
@naydav  makes sense, naming should be aligned to outline the best practices and make it understandable for external developers what's the responsibility of the class reading its name.

naydav commented on app/code/Magento/Inventory/Indexer/Scope/UnknownStateException.php:
```diff
@@ -0,0 +1,18 @@
+ * @todo refactoring it copy from catalog search module
+ * @api
+ */
+class UnknownStateException extends LocalizedException
```
Need to think about moving this exception to framework lib if we will use it

maghamed replied:
@naydav agree, already pointed that out for IndexTableNotExistException.
These should be framework level exceptions, we can't provide them to client because they provide details of how this functionality is implemented (leaky abstractions).

So, these exceptions are not successors of Localized

maghamed replied:
Btw, I don't know why we have this Exception in our codebase. 
So, from what I see we have a pre-defined finite list of states which could be switched in pre-defined order. (aka [State machine design pattern](https://en.wikipedia.org/wiki/State_pattern)) 

- our state object initially being initialized in 'REGULAR' state.
- after that, we've decided to move it to 'TEMPORARY' state while making re-indexation
- after re-indexation complete - we move the object to 'REGULAR' state again.

Is it correct? I suppose so, then why do we have a possibility that State object would be created in a wrong state? 
@larsroettig

naydav commented on app/code/Magento/Inventory/Indexer/StockItem.php:
```diff
@@ -0,0 +1,123 @@
+
+        foreach ($stocks as $stock) {
+            $stockId = $stock[StockInterface::STOCK_ID];
+            $dimensions = [$this->dimensionFactory->create(['name' => 'stock_', 'value' => $stockId])];
```
Looks like name should be `scope`
`'name' => 'scope'`

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        $select->group(['sku', 'status']);
+        return new \ArrayIterator($connection->fetchAll($select));
```
Why do we need to wrap the result in ArrayIterator object?
If we want to work with batch we need to return Generator... if I not mistake

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,81 @@
+     * @param \Magento\Framework\DB\Adapter\AdapterInterface $connection
+     * @return string
+     */
+    private function getSourceItemTableName($connection)
```
Looks like this method is redundant

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,81 @@
+     * @param \Magento\Framework\DB\Adapter\AdapterInterface $connection
+     * @return string
+     */
+    private function getLinkTableName($connection)
```
Looks like this method is redundant

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,81 @@
+{
+
+    const TABLE_NAME_SOURCE_ITEM = 'inventory_source_item';
+    const TABLE_NAME_STOCK_SOURCE_LINK = 'inventory_source_stock_link';
```
We already have these constants in our ResourceModel's classes

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/Dimension.php:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Stock Item Dimension
+ */
+class Dimension extends \Magento\Framework\Search\AbstractKeyValuePair
```
Need to think this class to framework 
@maghamed what do you think about this?

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,134 @@
+use Magento\Framework\Indexer\SaveHandler\Batch;
+use Magento\Framework\Indexer\SaveHandler\IndexerInterface as SaveHandlerIndexerInterface;
+use Magento\Framework\Search\Request\Dimension;
+use Magento\Framework\Search\Request\DimensionFactory;
```
Why do we use `Magento\Framework\Search\Request\DimensionFactory` instead of `Magento\Inventory\Indexer\StockItem\Dimension` declared in this PR ?

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,134 @@
+        IndexScopeResolverInterface $indexScopeResolver,
+        Batch $batch,
+        ResourceConnection $resource,
+        $batchSize = 100
```
Move this option to DI configuration

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * @var  ResourceConnection
+     */
+    private $resource;
```
Provide more clear name like `connection`

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/IndexHandler.php:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * @inheritdoc
+     */
+    public function deleteIndex($dimensions, \Traversable $documents)
```
Why are $documents `array` in one case but in another $documents are `\Traversable`

naydav commented on app/code/Magento/Inventory/Indexer/StockItem.php:
```diff
@@ -0,0 +1,123 @@
+     */
+    public function executeFull()
+    {
+        $stocks = $this->assignedStocksForSource->execute([]);
```
Looks like logic error

We work with `GetAssignedStocksInterface $assignedStocksForSource`, according to it interface returned value is `@return int[] List of stock ids`

So `foreach ($stocks as $stock) {` must be replaced on `foreach ($stockIds as $stockId) {`

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Stock Item Dimension
+ */
+class DataProvider
```
Is it index data builder? Maybe need to choose more clear name

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,81 @@
+        /** @var \Magento\Framework\DB\Adapter\AdapterInterface $connection */
+        $connection = $this->resource->getConnection();
+
+        $columns = ['sku', 'SUM(quantity) as quantity', 'status'];
```
`SUM(quantity)`
Why do we not consider that the status can be `disabled`?

naydav commented on app/code/Magento/Inventory/Indexer/StockItem.php:
```diff
@@ -0,0 +1,123 @@
+        $stocks = $this->assignedStocksForSource->execute($ids);
+
+        foreach ($stocks as $stockId) {
+            $dimensions = [$this->dimensionFactory->create(['name' => 'stock', 'value' => $stockId])];
```
Why do we use plural form?
We can work only with single dimension

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/DataProvider.php:
```diff
@@ -0,0 +1,81 @@
+            $select->where(SourceInterface::SOURCE_ID . ' (?)', $sourceIds);
+        }
+
+        $select->group(['sku', 'status']);
```
If the last record has `disabled` status then we obtain invalid index data (if table has records with both statuses)

naydav commented on app/code/Magento/Inventory/Model/GetAssignedStocksForSource.php:
```diff
@@ -0,0 +1,111 @@
+/**
+ * @inheritdoc
+ */
+class GetAssignedStocksForSource implements GetAssignedStocksForSourceInterface
```
Now we have `GetAssignedStocks`, so this class looks like redundant

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/Service/GetAssignedStocksInterface.php:
```diff
@@ -0,0 +1,19 @@
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Inventory\Indexer\StockItem\Service;
```
Pls remove namespace `Service`

naydav commented on app/code/Magento/Inventory/Indexer/StockItem/Service/GetAssignedStocks.php:
```diff
@@ -0,0 +1,46 @@
+            $select->where(SourceInterface::SOURCE_ID . ' (?)', $sourceIds);
+        }
+        $select->group(StockInterface::STOCK_ID);
+        return $connection->fetchAll($select);
```
We need to use `fetchCol`

naydav commented on app/code/Magento/Inventory/etc/di.xml:
```diff
@@ -64,5 +64,30 @@
     <!-- StockSourceLink -->
     <preference for="Magento\InventoryApi\Api\AssignSourcesToStockInterface" type="Magento\Inventory\Model\AssignSourcesToStock"/>
     <preference for="Magento\InventoryApi\Api\GetAssignedSourcesForStockInterface" type="Magento\Inventory\Model\GetAssignedSourcesForStock"/>
+    <preference for="Magento\InventoryApi\Api\GetAssignedStocksForSourceInterface" type="Magento\Inventory\Model\GetAssignedStocksForSource"/>
```
Now we have GetAssignedStocksInteface, so it looks like redundant

naydav commented on app/code/Magento/InventoryApi/Api/GetAssignedStocksForSourceInterface.php:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @api
+ */
+interface GetAssignedStocksForSourceInterface
```
Now we have GetAssignedStocksInteface, so it looks like redundant

naydav commented on composer.lock:
```diff
@@ -4,6 +4,7 @@
         "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file",
```
Why do we need to change composer.lock ?

naydav commented on dev/tests/integration/testsuite/Magento/Inventory/Indexer/StockItemTest.php:
```diff
@@ -0,0 +1,40 @@
+
+    /**
+     * @magentoDbIsolation disabled
+     * @magentoDataFixture Magento/Inventory/_files/products.php
```
Maybe it is better to use one fixture with all preconditions

naydav commented on dev/tests/integration/testsuite/Magento/Inventory/_files/products.php:
```diff
@@ -0,0 +1,66 @@
+    ->setStatus(\Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_ENABLED);
+
+/** @var \Magento\Catalog\Api\ProductRepositoryInterface $productRepositoryFactory */
+$productRepository = $objectManager->create(\Magento\Catalog\Api\ProductRepositoryInterface::class);
```
Why do you create an instance of repository each time?

naydav commented on dev/tests/integration/testsuite/Magento/Inventory/_files/products.php:
```diff
@@ -0,0 +1,66 @@
+ * See COPYING.txt for license details.
+ */
+
+\Magento\TestFramework\Helper\Bootstrap::getInstance()->reinitialize();
```
You can reuse fixture from catalog module

naydav commented on dev/tests/integration/testsuite/Magento/Inventory/_files/source.php:
```diff
@@ -0,0 +1,66 @@
+<?php
+/**
```
We already have this fixture
app/code/Magento/InventoryApi/Test/_files/source/source_list.php

naydav commented on dev/tests/integration/testsuite/Magento/Inventory/_files/stock.php:
```diff
@@ -0,0 +1,38 @@
+<?php
```
app/code/Magento/InventoryApi/Test/_files/source/stock_list.php

Commit:
MSI: WIP : Msi stock indexing

-- fixes after merge

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento-engcom/magento2?pullRequest=49) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento-engcom/magento2?pullRequest=49) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: larsroettig<br/>:x: naydav

maghamed commented on app/code/Magento/Inventory/Indexer/Scope/IndexTableNotExistException.php:
```diff
@@ -0,0 +1,19 @@
+ * @todo refactoring it copy from catalog search module
+ * @api
+ */
+class IndexTableNotExistException extends LocalizedException
```
If we are introducing IndexTableNotExistException, this exception should be low-level (framework level). 
Not inherited from LocalizedException.
Because we can't provide this Exception o the client code, even its name is a leaky abstraction of the internal details.

maghamed commented on app/code/Magento/Inventory/Indexer/Scope/IndexSwitcherInterface.php:
```diff
@@ -0,0 +1,24 @@
+<?php
```
In Magento there is another Switcher mechanism - https://github.com/magento-engcom/magento2/blob/develop/app/code/Magento/Catalog/Model/ResourceModel/Indexer/ActiveTableSwitcher.php

by our request, it would be moved to the Framework soon (in the scope of Magento 2.2.0 regression).
So, we could consider using ActiveTableSwitcher for our purposes.

@naydav please consider this and provide a solution for @larsroettig 

maybe we need some additional classes to be moved to the Framework?

maghamed reviewed (commented):
@naydav  please help @larsroettig 
and join this story development. 

It's very sophisticated Story, and there are no good examples in Magento where re-indexation was implemented totally correct. 

So, let's make a good example of introducing New Index to the Magento 2. Outlining all the steps and document them afterwards. 

We already agreed that ActiveTableSwitcher would be moved to Framework maybe we need some additional things to move to the framework level?

Commit:
MSI magento-engcom/magento#48:
- Add test for index row

Commit:
Merge remote-tracking branch 'origin/develop' into msi-stock-indexing

# Conflicts:
#	app/code/Magento/Inventory/etc/di.xml

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- refactoring

Commit:
WIP: Msi stock indexing #49

- fixes

Commit:
WIP: Msi stock indexing #49

- fixes

Commit:
WIP: Msi stock indexing #49

- fix static tests
- update db structure

Commit:
WIP: Msi stock indexing #49

- fix static tests

naydav commented:
Internal build
https://bamboo.corp.magento.com/browse/M2-AT2478-3

- Internal static build (INTEGRITY-SANITY-EE) is red due to MAGETWO-71596
- Internal static build (Static LEGACY-EE-L64 and EE-L64) is red. 
But coupling between objects for `\Magento\Inventory\Test\Integration\Indexer\StockItemTest` will be fixed during https://github.com/magento-engcom/magento2/issues/69

maghamed commented:
Approved to be merged into the Develop branch 

Great job guys @larsroettig  @naydav , well done!
